### PR TITLE
datapath: Do not use ENABLE_HOST_SERVICES_FULL

### PR DESCRIPTION
**Pull request by jrajahalme**
Move logic used to define ENABLE_HOST_SERVICES_FULL to datapath. This
fixes verifier complexity tests that defined all the right flags
(-DENABLE_HOST_SERVICES_TCP=1 -DENABLE_HOST_SERVICES_UDP=1), but did not
define ENABLE_HOST_SERVICES_FULL, so the datapath code still pulled in
the service lookup code in test-verifier test cases.

Host services are not supported in kernel 4.9
(https://docs.cilium.io/en/latest/gettingstarted/host-services/), so
remove them from verifier tests for 4.9. This actually reverts the lack
of ENABLE_HOST_SERVICES_FULL definition and causes the per-packet lb code
to be pulled in to bpf_lxc just like before this patch when
ENABLE_HOST_SERVICES_TCP ENABLE_HOST_SERVICES_UDP were defined.

For the other kernel versions add test cases for bpf_lxc where host
services are not available, as that stresses another datapath
configuration (bpf_lxc uses ENABLE_PER_PACKET_LB in this case).

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

### CODE REVIEW

**Comment by jrajahalme** [comment]
runtime test hit known flake: #19124

**Comment by jrajahalme** [comment]
gke hit by an old flake #6128

**Comment by jrajahalme** [comment]
/test-1.23-net-next

**Review by ldelossa — APPROVED** [review]
LGTM, did a sanitcy check on config_test.go and all looks well.

File: bpf/bpf_lxc.c [hunk 1]
```diff
@@ -51,8 +51,10 @@
 #include "lib/nodeport.h"
 #include "lib/policy_log.h"
 
-#if !defined(ENABLE_HOST_SERVICES_FULL) || defined(ENABLE_SOCKET_LB_HOST_ONLY)
-# define ENABLE_PER_PACKET_LB
+#if !defined(ENABLE_HOST_SERVICES_TCP) || \
+  !defined(ENABLE_HOST_SERVICES_TCP) || \
```
Comment (borkmann) [inline bpf/bpf_lxc.c]: `!defined(ENABLE_HOST_SERVICES_TCP)` should this say `!defined(ENABLE_HOST_SERVICES_UDP)` otherwise it's a dup?

**Review by borkmann — CHANGES_REQUESTED** [review]

File: bpf/bpf_lxc.c (hunk 1 above)
Comment (jrajahalme) [reply bpf/bpf_lxc.c]: Yes, it was a copy&paste error, thanks for noticing!

**Comment by jrajahalme** [comment]
Fixed copy&paste error and rebased to master changes.

**Review by borkmann — APPROVED** [review]

**Comment by jrajahalme** [comment]
Flake #18912 in test-gke

**Comment by jrajahalme** [comment]
GKE flake #19122

**Comment by jrajahalme** [comment]
net-next fail #19264: https://jenkins.cilium.io/job/Cilium-PR-K8s-1.23-kernel-net-next/873/testReport/junit/Suite-k8s-1/23/K8sBGPTests_Tests_LoadBalancer_Connectivity_to_endpoint_via_LB/

**Comment by jrajahalme** [comment]
/test-1.23-net-next

**Comment by jrajahalme** [comment]
/test-1.23-net-next

**Comment by jrajahalme** [comment]
/test-1.23-net-next

**Comment by jrajahalme** [comment]
/test-1.23-net-next

**Comment by jrajahalme** [comment]
Rebased to fix merge conflict with master

**Review by pchaigno — CHANGES_REQUESTED** [review]
> This
fixes verifier complexity tests that defined all the right flags
(-DENABLE_HOST_SERVICES_TCP=1 -DENABLE_HOST_SERVICES_UDP=1), but did not
define ENABLE_HOST_SERVICES_FULL, so the datapath code still pulled in
the service lookup code in test-verifier test cases.

That was on purpose. KPR + per-packet LB is possible since recently, so we want to test the configuration that includes those two features, as that stresses complexity the most.

**Comment by jrajahalme** [comment]
> > This
> > fixes verifier complexity tests that defined all the right flags
> > (-DENABLE_HOST_SERVICES_TCP=1 -DENABLE_HOST_SERVICES_UDP=1), but did not
> > define ENABLE_HOST_SERVICES_FULL, so the datapath code still pulled in
> > the service lookup code in test-verifier test cases.
> 
> That was on purpose. KPR + per-packet LB is possible since recently, so we want to test the configuration that includes those two features, as that stresses complexity the most.

For kernels > 4.9 existing complexity tests are kept. For 4.9 I removed `DENABLE_HOST_SERVICES_TCP` and `ENABLE_HOST_SERVICES_UDP` as I understood them requiring a newer kernel, is that not the case?

Also, since per-packet-lb now involves a new tailcall it seems like a good idea to test the complexity with and without it, no?

**Comment by jrajahalme** [comment]
/test-1.23-net-next

**Comment by jrajahalme** [comment]
> > This
> > fixes verifier complexity tests that defined all the right flags
> > (-DENABLE_HOST_SERVICES_TCP=1 -DENABLE_HOST_SERVICES_UDP=1), but did not
> > define ENABLE_HOST_SERVICES_FULL, so the datapath code still pulled in
> > the service lookup code in test-verifier test cases.
> 
> That was on purpose. KPR + per-packet LB is possible since recently, so we want to test the configuration that includes those two features, as that stresses complexity the most.

@pchaigno Maybe defining `ENABLE_SOCKET_LB_HOST_ONLY` or `ENABLE_L7_LB` in addition to `ENABLE_HOST_SERVICES_TCP` and `ENABLE_HOST_SERVICES_UDP` in the complexity tests would be a more explicit way of testing this combination? Do you have a preference?

**Comment by pchaigno** [comment]
> For kernels > 4.9 existing complexity tests are kept. For 4.9 I removed `DENABLE_HOST_SERVICES_TCP` and `ENABLE_HOST_SERVICES_UDP` as I understood them requiring a newer kernel, is that not the case?

I'm fine with the changes, but the commit description needs to be changed :-)

> Also, since per-packet-lb now involves a new tailcall it seems like a good idea to test the complexity with and without it, no?

Yep, that's a good point :+1: 

> Maybe defining `ENABLE_SOCKET_LB_HOST_ONLY` or `ENABLE_L7_LB` in addition to `ENABLE_HOST_SERVICES_TCP` and `ENABLE_HOST_SERVICES_UDP` in the complexity tests would be a more explicit way of testing this combination? Do you have a preference?

`ENABLE_SOCKET_LB_HOST_ONLY` sounds good, but I'm also fine with the current approach.

**Comment by jrajahalme** [comment]
@pchaigno Updated the commit message, see if it is satisfactory :-)

**Comment by aanm** [comment]
@pchaigno one of the tests failed (`K8sVerifier Runs the kernel verifier against Cilium's BPF datapath`), I guess that's good news?

File: bpf/complexity-tests/49/bpf_host.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
...
+-DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 -DENABLE_HOST_REDIRECT=1 -DENABLE_ROUTING=1 -DNO_REDIRECT=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_IPSEC=1 -DIP_POOLS=1 -DENCAP_IFINDEX=1 -DTUNNEL_MODE=1
+-DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_HOST_REDIRECT=1 -DENABLE_ROUTING=1 -DNO_REDIRECT=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_IPSEC=1 -DIP_POOLS=1 -DENCAP_IFINDEX=1 -DTUNNEL_MODE=1
+-DSKIP_DEBUG=1 -DENABLE_IPV6=1 -DENABLE_HOST_REDIRECT=1 -DENABLE_ROUTING=1 -DNO_REDIRECT=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_IPSEC=1 -DIP_POOLS=1 -DENCAP_IFINDEX=1 -DTUNNEL_MODE=1
+
```
Comment (pchaigno) [inline bpf/complexity-tests/49/bpf_host.txt:4]: This empty line is most likely why CI is failing.

Comment (pchaigno) [reply bpf/complexity-tests/49/bpf_host.txt:4]: Same for the other empty lines below.

**Review by pchaigno — CHANGES_REQUESTED** [review]

**Comment by brb** [comment]
Are we going to backport this to v1.12 (or older)? If not, then I suggest to close this PR, as in v1.13 we are removing the host-reachable svc protocols (and thus `ENABLE_HOST_SERVICES_FULL`).

**Pull request closed without merging**